core/server/controllers/frontend/templates.js

```diff
@@ -74,17 +74,19 @@ function getChannelTemplateHierarchy(channelOpts) {
  */
 function getSingleTemplateHierarchy(postObject) {
     var templateList = ['post'],
-        type = 'post';
+        slugTemplate = 'post-' + postObject.slug;
 
     if (postObject.page) {
         templateList.unshift('page');
-        type = 'page';
+        slugTemplate = 'page-' + postObject.slug;
     }
 
-    templateList.unshift(type + '-' + postObject.slug);
+    templateList.unshift(slugTemplate);
 
     if (postObject.custom_template) {
-        templateList.unshift(postObject.custom_template);
+        if (themes.getActive() && !themes.getActive().hasTemplate(slugTemplate)) {
+            templateList.unshift(postObject.custom_template);
+        }
```

Comment:
I think this logic is not quite right. The template hierarchy is a list in order of preference.

The doc block says the list is: `custom-*, post-:slug, post`, but that is not correct!

The list should be `post-:slug, custom-*, post`, I think?

If there is an explicit matching template, that should be used first, then the custom template, if provided, and then post.hbs as the fallback.

I think moving the `if (postObject.custom_template)` statement above the  `templateList.unshift(slugTemplate);` line would fix the logic?

Reply:
The actual end result is correct, because we check if the slug template exists. See [tests](https://github.com/TryGhost/Ghost/pull/9073/files#diff-eccd2c32973b912b29a55a0bd382e25b).

But agree, let's improve the readability here, can be simplified by re-ordering the array shifts.

Reply:
https://github.com/TryGhost/Ghost/pull/9073/commits/5c21605120ef6c438b2504f12d15e54ea7545ae2